Assign priorities to configuration scopes

alalazo opened the pull request:
fixes #48414

Currently, environments can end up with higher priority than `-C` custom config scopes and `-c` command line arguments sometimes. This shouldn't happen -- those explicit CLI scopes should override active environments.

Up to now configuration behaved like a stack, where scopes could be only be pushed at the top. This PR allows to assign priorities to scopes, and ensures that scopes of lower priorities are always "below" scopes of higher priorities.

When scopes have the same priority, what matters is the insertion order.

Modifications:
- [x] Add a mapping that iterates over keys according to priorities set when adding the key/value pair
- [x] Use that mapping to allow assigning priorities to configuration scopes
- [x] Assign different priorities for different kind of scopes, to fix a bug, and add a regression test
- [x] Simplify `Configuration` constructor f1d03715fda47d6f498ca1838911e0a987af2888
- [x] Remove `Configuration.pop_scope` 61dd2f291db3ff409220e9f2dfb80a2c7608bb28
- [x] Remove `unify:false` from custom `-C` scope in pipelines 4a5a12dcd9c7612a1f325e37cdeed4492a53b4c8 On `develop`, pipelines are relying on the environment being able to override `-C` scopes, which is a bug. After this fix, we need to be explicit about the unification strategy in each stack, and remove the blanket unify:false from the highest priority scope
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

alalazo committed:
Add PriorityOrderedMapping

alalazo committed:
Use PriorityOrderedMapping instead of collections.OrderedDict for config

alalazo committed:
Add an explicit "priority" arg to Configuration.push_scope

alalazo committed:
Assign priorities to config scopes

Up to now configuration behaved like a stack, where scopes could be
only be pushed at the top.

This commit allows to assign priorities to scopes, and ensures that
scopes of lower priorities are always "below" scopes of higher
priorities.

When scopes have the same priority, what matters is the insertion order.

alalazo committed:
Simplify Configuration constructor

alalazo committed:
Remove Configuration.pop_scope

The function is only used in a fixture, and it can be substituted
with already existing code.

alalazo committed:
Slightly improve docstrings

alalazo committed:
Import Mapping from typing

alalazo committed:
Add exceptions to docs

alalazo committed:
Add exceptions to docs

alalazo committed:
Fix docstring

alalazo committed:
Remove unneeded enum value

alalazo commented:
Failure in pipelines is unrelated, and due to system errors:
- https://gitlab.spack.io/spack/spack/-/pipelines/942431
- https://gitlab.spack.io/spack/spack/-/pipelines/942475

haampie commented on lib/spack/llnl/util/lang.py:
```diff
@@ -1080,3 +1080,104 @@ def __set__(self, instance, value):
+        if key in self._data:
+            self.remove(key)
+
+        self._priorities.append((priority, self._new_id(), key))
```
`sort`/`sorted` is stable, so no need for `_new_id()`.

alalazo replied:
~We need that~ It's useful to be defensive when iterating (in case a new function scrambles the order of `_priorities`). 

EDIT: I changed that in https://github.com/spack/spack/pull/48420/commits/87752f1e5a44abb7074d189be397a556291dc8c6 This is more similar to my first implementation, which I modified by adding `_new_id()` out of possibly an abundance of caution - motivated by the numerous bugs we had with config scopes order. If we have a second thought, and want to get back to having an explicit index, we can drop the commit.

alalazo committed:
Rely on sort being stable, and on _priorities being ordered correctly

psakievich reviewed (dismissed):
I've looked over the code and the concept seems sound to me. The idea of stack like behavior within a given priority is a more robust design. It also seems truer to our description of how configs will work with the users. I'm still mulling a bit about the defaults selected but not to the point of wanting to hold up a merge. Thanks @alalazo for taking the time to shore up this design.

haampie commented on lib/spack/llnl/util/lang.py:
```diff
@@ -1080,3 +1080,101 @@ def __set__(self, instance, value):
+        return len(self._data)
+
+    def __iter__(self):
+        def iter_fn():
+            for *_, key in self._priorities:
+                yield key
+
+        return iter_fn()
```
```suggestion
        yield from (key for _, key in self._priorities)
```

haampie commented on lib/spack/llnl/util/lang.py:
```diff
@@ -1080,3 +1080,101 @@ def __set__(self, instance, value):
+        return iter_fn()
+
+    def __reversed__(self):
+        def iter_fn():
+            for *_, key in reversed(self._priorities):
+                yield key
+
+        return iter_fn()
```
```suggestion
        yield from (key for _, key in reversed(self._priorities))
```

haampie commented on lib/spack/llnl/util/lang.py:
```diff
@@ -1080,3 +1080,101 @@ def __set__(self, instance, value):
+
+    def reversed_values(self):
+        """Iterates over values from the highest priority, to the lowest."""
+
+        def iter_fn():
+            for *_, key in reversed(self._priorities):
+                yield self._data[key]
+
+        return iter_fn()
```
```suggestion
        yield from (self._data[key] for _, key in reversed(self._priorities))
```

haampie commented on lib/spack/llnl/util/lang.py:
```diff
@@ -1080,3 +1080,101 @@ def __set__(self, instance, value):
+    def _highest_priority(self) -> int:
+        if not self._priorities:
+            return 0
+        result, *_ = self._priorities[-1]
```
```suggestion
        result, _ = self._priorities[-1]
```

haampie commented:
I don't see any code change in bootstrap :thinking:  what about

```
spack -c config:install_tree:root:/tmp spec zlib
```

bootstrapping to /tmp instead of ~/.spack/bootstrap?

alalazo commented:
@haampie There shouldn't be any change to bootstrap. The bootstrap store is specified in another place, so if you want to modify it from the command line you should:
```console
$ spack -c bootstrap:root:/tmp spec zlib
```
even though the simpler way to do it is:
```console
$ spack bootstrap root /tmp
```
Anyhow, I'm not aiming at changing any bootstrap semantic in this PR.

alalazo committed:
Save a few lines

wdconinc commented:
@haampie I think this PR is approved by reviewers, but there are some unresolved conversations by you. Just wanna make sure it didn't fall off the radar.

alalazo commented:
@wdconinc This was supposed to wait a bit, so #46792 (which later became #48784) could be merged. I think it's now clear that the blocking PR needs more time to be merged than originally foreseen.

Also, I need to "fix" hep pipelines generation here. I see that on develop we want to have:
```
unify: when_possible
```
Unfortunately, we rely on the bug being fixed to get it, so here it seems we are using `unify: false`.

alalazo committed:
pipelines: make unification strategy explicit

This commit removes the "unify:false" in higher
priority custom scopes, and makes the unification
strategy explicit in each pipeline.

wdconinc commented:
> Also, I now see I need to "fix" hep pipelines generation here.

@alalazo If you want, we can switch the hep pipeline to `unify: false` as well. There are several expensive builds in there, though, and I'd want to make sure we don't build them more than necessary. If that's a direction to go in, I can do that in a separate PR. Let me know.

Alternatively, I was trying to move those hep pipelines closer to `unify: true`, so maybe that's a direction to go in as well.

alalazo commented:
@wdconinc I just pushed the fix in the last commit, just waiting to see pipelines :heavy_check_mark:

psakievich commented:
So are we waiting on https://github.com/spack/spack/pull/48784 or not? I guess I'm a little lost on that. I'll just keep approving for now...😅

tgamblin commented:
@kwryankrattiger see the pipeline changes here -- I think they'll be less relevant if we use less `-C` and more `include:`, but they're relevant at least with this change.

kwryankrattiger commented:
> @kwryankrattiger see the pipeline changes here -- I think they'll be less relevant if we use less `-C` and more `include:`, but they're relevant at least with this change.

It has been on my list to convert all of the stacks to include configs instead of using -C. It has been a source of more confusion than convenience and the original idea for dynamic scopes has proven to be better expressed through includes or using "config add".
